cmd/internal/cli/actions.go

```diff
@@ -217,6 +184,33 @@ func replaceURIWithImage(cmd *cobra.Command, args []string) {
 	return
 }
 
+// TODO: Let's stick this in another file so that that CLI is just CLI
+func vmStarter(cmd *cobra.Command, image string, args []string) bool {
+	// check if --vm-ram or --vm-cpu changed from default value
+	for _, flagName := range []string{"vm-ram", "vm-cpu"} {
+		if flag := cmd.Flag(flagName); flag != nil && flag.Changed {
+			// the user passed an option related to the VM...
+			if !VM {
+				// ... but didn't enable the VM itself
+				cmd.Flag("vm").Value.Set("true")
+			}
+			break
+		}
+	}
+
+	// since --syos is a boolean, it cannot be added to the above list
+	if IsSyOS && !VM {
+		sylog.Warningf("The --syos option requires a virtual machine, automatically enabling --vm option.")
+		cmd.Flag("vm").Value.Set("true")
+	}
+
+	if VM {
+		prepareVM(cmd, image, args)
+	}
+
+	return VM
+}
+
 // ExecCmd represents the exec command
 var ExecCmd = &cobra.Command{
```

Comment:
Following from the other comments, I think that the run/exec/shell Command variables should all have their `Run` fields be a function like this:

```
func ActionRun(cmd *cobra.Command, args []string) {
	a := append([]string{"/.singularity.d/actions/exec"}, args[1:]...)
	for _, flagName := range []string{"vm-ram", "vm-cpu"} {
		if flag := cmd.Flag(flagName); flag != nil && flag.Changed {
			// the user passed an option related to the VM...
			if !VM {
				// ... but didn't enable the VM itself
				cmd.Flag("vm").Value.Set("true")
			}
			break
		}
	}

	// since --syos is a boolean, it cannot be added to the above list
	if IsSyOS && !VM {
		sylog.Warningf("The --syos option requires a virtual machine, automatically enabling --vm option.")
		cmd.Flag("vm").Value.Set("true")
	}

	if VM {
		execVM()
	}

	execStarter()
}
```

On both `darwin` && `linux`, `execVM` will be defined (in fact they will both be defined here in `actions.go`). However, on all other platforms `execStarter` will be defined as `panic()`.